src/main/java/fi/aalto/cs/apluscourses/model/Module.java

```diff
@@ -0,0 +1,103 @@
+package fi.aalto.cs.apluscourses.model;
+
+import fi.aalto.cs.apluscourses.utils.Event;
+import fi.aalto.cs.apluscourses.utils.StateMonitor;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Collections;
+import java.util.List;
+import org.jetbrains.annotations.NotNull;
+import org.json.JSONObject;
+
+public class Module {
+
+  public static final int ERROR = StateMonitor.ERROR;
+  public static final int NOT_INSTALLED = StateMonitor.INITIAL;
+  public static final int FETCHING = NOT_INSTALLED + 1;
+  public static final int FETCHED = FETCHING + 1;
+  public static final int LOADING = FETCHED + 1;
+  public static final int LOADED = LOADING + 1;
+  public static final int WAITING_FOR_DEPS = LOADED + 1;
+  public static final int INSTALLED = WAITING_FOR_DEPS + 1;
+
+  public final Event stateChanged = new Event();
+  public final StateMonitor stateMonitor = new StateMonitor(this::onStateChanged);
+
+  @NotNull
+  private final String name;
+  @NotNull
+  private final URL url;
+
+  /**
+   * Constructs a module with the given name and URL.
+   * @param name The name of the module.
+   * @param url The URL from which the module can be downloaded.
+   */
+  public Module(@NotNull String name, @NotNull URL url) {
+    this.name = name;
+    this.url = url;
+  }
+
+  /**
+   * Returns a module constructed from the given JSON object. The object should contain the name of
+   * the module and the URL from which the module can be downloaded. Additional members in the
+   * object don't cause any errors. Example of a valid JSON object:
+   * <pre>
+   * {
+   *   "name": "My Module",
+   *   "url": "https://example.com"
+   * }
+   * </pre>
+   * @param jsonObject The JSON object containing information about a single module.
+   * @return A module constructed from the given JSON object.
+   * @throws MalformedURLException  If the URL of the module is malformed.
+   * @throws org.json.JSONException If the jsonObject doesn't contain "name" and "url" keys with
+   *                                string values.
+   */
+  @NotNull
+  public static Module fromJsonObject(@NotNull JSONObject jsonObject, ModelFactory factory)
+      throws MalformedURLException {
```

Comment:
Nit:
this is a bit controversial, as `JSONException` just happens to extend `RuntimeException`, but maybe it will improve the readability, if we at least add it to the `throw` clause like `throws MalformedURLException, JSONException`, thus, it does not really propagate further (as in [here](https://github.com/Aalto-LeTech/intellij-plugin/blob/e836952634484c6a0816cba4318a5a9c6d19d0c4/src/main/java/fi/aalto/cs/apluscourses/model/Course.java#L214), you just need to know, that the `JSONException` is coming). But this is tested, so I don't know...

Reply:
Sonar says `RuntimeException`s (like `JSONException`) should not be declared in `throws` statement. We certainly can mutiny against Sonar here, if you like, but I cannot say which way is better.

Reply:
If we want to have it in the `throws` clause, then I think we should catch `JSONException` here and throw our own exception that doesn't extend `RuntimeException`.

Reply:
I'll leave this for you, @nikke234, to decide as you are the original author of this method, right? In any case, I'd leave it out of this PR.